BFAIR/normalization/normalization.py

```diff
@@ -0,0 +1,364 @@
+        normalized, defaults to 'Intensity'
+    groupname_colname: string
+        the name of the column with the sample group names,
+        defaults to 'sample_group_name'
+
+    Returns
+    -------
+    output_df: pandas.DataFrame
+        the output dataframe. It follows the same architecture as
+        the input dataframe, just with normalized values
+    """
+    output_df = pd.DataFrame()
+    sample_group_names = df[groupname_colname].unique()
+    for i, sample_group_name in enumerate(sample_group_names):
+        new_df = copy.deepcopy(df[df[groupname_colname] == sample_group_name])
+        tsi = new_df[columnname].sum(skipna=True)
+        new_df[columnname] = new_df[columnname].div(tsi)
+        if i == 0:
+            output_df = new_df
+        else:
+            output_df = output_df.append(new_df)
+    return output_df
+
+
+def biomass_tsi_norm(
+    biomass_substrate_df,
+    df,
+    columnname="Intensity",
+    groupname_colname="sample_group_name",
+):
+    """
+    Applies a modified version of Total Sum Intensity normalization; all
+    values will be divided by the sum of all all the values of metabolites
+    that are part of the biomass function of a model of the organism of
+    interest. Like that, summing all the values for the metabolites
+    contributing to the biomass function up would amout to '1'
+
+    Parameters
+    ----------
+    biomass_substrate_df: pandas.DataFrame
+        a dataframe with one column denoting the
+        metabolites that are part of the biomass function and one
+        column with their corresponding multipliers
+    df: pandas.DataFrame
+        input dataframe, output of either the extractNamesAndIntensities()
+        or the calculateMeanVarRSD() function
+    columnname: string
+        the name of the column with the data that needs to be
+        normalized, defaults to 'Intensity'
+    groupname_colname: string
+        the name of the column with the sample group names,
+        defaults to 'sample_group_name'
+
+    Returns
+    -------
+    output_df: pandas.DataFrame
+        the output dataframe. It follows the same architecture as
+        the input dataframe, just with normalized values
+    """
+    bm_tsi = 0
+    output_df = pd.DataFrame()
+    sample_group_names = df[groupname_colname].unique()
+    for i, sample_group_name in enumerate(sample_group_names):
+        new_df = copy.deepcopy(df[df[groupname_colname] == sample_group_name])
+        for biomass_met in biomass_substrate_df["Metabolite"]:
+            met_tsi = sum(
+                new_df[new_df["Metabolite"] == biomass_met][columnname]
+            )
+            bm_tsi += met_tsi
+        norm_vals = new_df[columnname].div(bm_tsi)
+        new_df[columnname] = norm_vals
+        if i == 0:
+            output_df = new_df
+        else:
+            output_df = output_df.append(new_df)
+    return output_df
+
+
+def biomass_formula_tsi_norm(
```

Comment:
This method also includes the stoichiometric coefficients of the biomass function (and the product which I left out in the previous function because I thought that other things on the product side can be neglected). This makes only a small difference but it might be interesting